src/app.ts

```diff
@@ -1,43 +1,76 @@
+import { BigNumberish } from 'ethers';
+
 import {
   ChainName,
-  HyperlaneApp,
-  objMap,
-  promiseObjAll,
+  ChainNameToDomainId,
+  GasRouterApp,
+  RouterContracts,
 } from '@hyperlane-xyz/sdk';
+import { Address } from '@hyperlane-xyz/utils/dist/src/types';
 
 import { HypERC20Contracts, HypERC721Contracts } from './contracts';
+import { TokenRouter } from './types';
 
-export class HypERC20App<Chain extends ChainName> extends HyperlaneApp<
+class HyperlaneTokenApp<
+  Contracts extends RouterContracts<TokenRouter>,
+  Chain extends ChainName,
+> extends GasRouterApp<Contracts, Chain> {
+  async transfer<Origin extends Chain>(
+    origin: Origin,
+    destination: Exclude<Chain, Origin>,
+    recipient: Address,
+    amountOrId: BigNumberish,
+  ) {
+    const originRouter = this.getContracts(origin).router;
+    const destinationDomain = ChainNameToDomainId[destination];
+    const gasPayment = await originRouter.quoteGasPayment(destinationDomain);
+    const chainConnection = this.multiProvider.getChainConnection(origin);
+    return chainConnection.handleTx(
+      originRouter.transferRemote(destinationDomain, recipient, amountOrId, {
+        value: gasPayment,
+      }),
+    );
+  }
+}
+
+export class HypERC20App<Chain extends ChainName> extends HyperlaneTokenApp<
   HypERC20Contracts,
   Chain
 > {
-  getSecurityModules = () =>
-    promiseObjAll(
-      objMap(this.contractsMap, (_, contracts) =>
-        contracts.router.interchainSecurityModule(),
-      ),
-    );
-
-  getOwners = () =>
-    promiseObjAll(
-      objMap(this.contractsMap, (_, contracts) => contracts.router.owner()),
-    );
+  async transfer<Origin extends Chain>(
+    origin: Origin,
+    destination: Exclude<Chain, Origin>,
+    recipient: Address,
+    amount: BigNumberish,
+  ) {
+    const originRouter = this.getContracts(origin).router;
+    const chainConnection = this.multiProvider.getChainConnection(origin);
+    const signerAddress = await chainConnection.signer!.getAddress();
+    const balance = await originRouter.balanceOf(signerAddress);
+    if (balance.lt(amount))
+      throw new Error(
```

Comment:
Should this happen here really? Idk what other higher-level tooling does, but in the case of a safe, you can still create the tx in the safe even if the safe currently does not have any balance right?